server/DataStorageController.js

```diff
@@ -15,142 +15,147 @@ const password = process.env.DB_PASSWORD;
-				if (err) throw err;
-				categoryList = result;
-			});
+function getCategories () {
+	pool.query('SELECT * FROM categories', async function(err, result, fields) {
+		if (err) {
+			handleEmptyDBError()
+			return getCategories()
 		}
-		else
-		{
-			categoryList = result;
+		else {
+			categoryList = await result
 		}
-	});
+	})
 }
 
-function getQuestions(){
-	con.query("SELECT * FROM questions", function (err, result, fields){
-		if(err)
-		{
-			con.query("DROP TABLE IF EXISTS questions", function (err, result){
-				if (err) throw err;
-			});
-			con.query("DROP TABLE IF EXISTS categories", function (err, result){
-				if (err) throw err;
-			});
-
-			// Empty or bad database
-			sqldumpImporter();
-
-			con.query("SELECT * FROM questions", function (err, result, fields){
-				if (err) throw err;
-				questionList = result;
-			});
+function getQuestions () {
+	pool.query("SELECT * FROM questions", async function (err, result, fields) {
+		if (err) {
+			handleEmptyDBError()
+			return getQuestions()
 		}
-		else
-		{
-			questionList = result;
+		else {
+			questionList = await result
 		}
-	});
+	})
 }
 
-var questionList;
-getQuestions();
-var categoryList;
-getCategories();
+function handleEmptyDBError () {
+	pool.query("DROP TABLE IF EXISTS questions", function (err, result) {
+		if (err) {
+			console.error(err)
+		}
+	})
+	pool.query("DROP TABLE IF EXISTS categories", function (err, result) {
+		if (err) {
+			console.error(err)
+		}
+	})
 
-// Exports
+	// Empty or bad database
+	sqldumpImporter()
+}
 
-module.exports.readAQuestion = function(){
-	return questionList[Math.floor(Math.random() * questionList.length)];
+function sleep(ms) {
+	return new Promise(resolve => setTimeout(resolve, ms))
 }
 
-module.exports.updateCategoryName = function(categoryID, newCategoryName){
-	con.query("UPDATE categories SET name = '" + newCategoryName + "' WHERE id = " + categoryID + ";", function (err, result){
-		if (err) throw err;
-		sqldumpexporter();
-		getCategories();
-	});
+let questionList
+getQuestions()
+let categoryList
+getCategories()
+
+// Exports --------------------------------------------------------------------
+module.exports.readAQuestion = function () {
+	return questionList[Math.floor(Math.random() * questionList.length)]
 }
 
-module.exports.updateCategoryColor = function(categoryID, newCategoryColor){
-	con.query("UPDATE categories SET color = '" + newCategoryColor + "' WHERE id = " + categoryID + ";", function (err, result){
-		if (err) throw err;
-		sqldumpexporter();
-		getQuestions();
-	});
+module.exports.updateCategories = async function(categories) {
+	for (c in categories) {
+		const category = categories[c]
+		pool.query(
+			`
+				UPDATE categories
+				SET name = '${category.name}', color = '${category.color}'
+				WHERE id = ${category.id};
+			`, function(err, result) {
+				if (err) console.error(err)
+			}
+		)
+	}
+
+	// DB doesn't have enough time to commit w/o this sleep here
+	await sleep(300)
```

Comment:
This sleep seems like it could cause inconsistencies maybe in the future. Do you think it is safe to assume it always be this amount of time?

Reply:
Most likely. I'll be writing the tests in the next few days so if I run into any issues I'll fix it.